pep-0999.rst

```diff
@@ -0,0 +1,541 @@
+recommended that the ``TYPE|None`` syntax be preferred over
+``Optional[TYPE]`` for marking such item values, to avoid using
+``Required[...]`` or ``NotRequired[...]`` alongside ``Optional[...]``
+within the same TypedDict definition:
+
+Yes:
+
+::
+
+   from __future__ import annotations  # for Python 3.7-3.9
+
+   class Dog(TypedDict):
+       name: str
+       owner: NotRequired[str|None]
+
+Avoid (unless Python 3.5-3.6):
+
+::
+
+   class Dog(TypedDict):
+       name: str
+       # ick; avoid using both Optional and NotRequired
+       owner: NotRequired[Optional[str]]
+
+
+Reference Implementation
+========================
+
+The following will be true when
+`mypy#9867 <https://github.com/python/mypy/issues/9867>`__ is
+implemented:
+
+    The `mypy <http://www.mypy-lang.org/>`__ type checker supports
+    ``Required`` and ``NotRequired``. A reference implementation of the
+    runtime component is provided in the
+    `typing_extensions <https://github.com/python/typing/tree/master/typing_extensions>`__
+    module.
+
+
+Rejected Ideas
+==============
+
+Special syntax around the *key* of a TypedDict item
+---------------------------------------------------
+
+::
+
+   class MyThing(TypedDict):
+       opt1?: str  # may not exist, but if exists, value is string
+       opt2: Optional[str]  # always exists, but may have null value
+
+or:
+
+::
+
+   class MyThing(TypedDict):
+       Optional[opt1]: str  # may not exist, but if exists, value is string
+       opt2: Optional[str]  # always exists, but may have null value
+
+These syntaxes would require Python grammar changes and it is not
+believed that marking TypedDict items as required or potentially-missing
+would meet the high bar required to make such grammar changes.
+
+Also, “let’s just not put funny syntax before the colon.” [1]_
+
+
+Marking required or potentially-missing keys with an operator
+-------------------------------------------------------------
+
+We could use unary ``+`` as shorthand to mark a required key, unary
+``-`` to mark a potentially-missing key, or unary ``~`` to mark a key
+with opposite-of-normal totality:
+
+::
+
+   class MyThing(TypedDict, total=False):
+       req1: +int    # + means a required key, or Required[...]
+       opt1: str
+       req2: +float
+
+   class MyThing(TypedDict):
+       req1: int
+       opt1: -str    # - means a potentially-missing key, or NotRequired[...]
+       req2: float
+
+   class MyThing(TypedDict):
+       req1: int
+       opt1: ~str    # ~ means a opposite-of-normal-totality key
+       req2: float
+
+Such operators could be implemented on ``type`` via the ``__pos__``,
+``__neg__`` and ``__invert__`` special methods without modifying the
+grammar.
+
+It was decided that it would be prudent to introduce longform syntax
+(i.e. ``Required[...]`` and ``NotRequired[...]``) before introducing
+shortform syntax. Future PEPs may reconsider introducing these shortform
+syntax options.
```

Comment:
In addition, these operators are pretty arbitrary. People have proposed unary `+` and `-` for various other type system features (e.g. to indicate covariance/contravariance).

Reply:
Did reword to make it more clear that this is just a syntax _idea_ and not necessarily the one we'll go with in the future. Now reads as:

```
It was decided that it would be prudent to introduce longform syntax
(i.e. ``Required[...]`` and ``NotRequired[...]``) before introducing
any shortform syntax. Future PEPs may reconsider introducing this
or other shortform syntax options.
```